File: packages/gbnf-experiment/data/20260909T194429Z_a7bb3ecf/ported_implementation/grammar_graph/generic_set.py
from __future__ import annotations

from typing import Callable, Dict, Generic, Iterator, Optional, TypeVar

T = TypeVar('T')
K = TypeVar('K')


class GenericSet(Generic[T, K]):
    """A set of elements deduplicated by a derived key.

    The reference keeps a `Map` of keys plus a `Set` of the elements; the `Set`
    uses reference identity, which is reproduced here by keying on `id()`.
    """

    def __init__(self, get_key: Callable[[T], K]):
        self.get_key = get_key
        self._keys: Dict[K, T] = {}
        self._set: Dict[int, T] = {}

    def add(self, el: T) -> None:
        key = self.get_key(el)
        if key not in self._keys:
            self._keys[key] = el
            self._set[id(el)] = el

    def delete(self, el: T) -> None:
        key = self.get_key(el)
        ref = self._keys.get(key)
        if ref is None:
            raise Exception('Could not get ref')
        del self._keys[key]
        self._set.pop(id(ref), None)

    def has(self, el: T) -> bool:
        return id(el) in self._set

    def get(self, el: T) -> Optional[T]:
        return self._keys.get(self.get_key(el))

    def __iter__(self) -> Iterator[T]:
        yield from list(self._set.values())

    @property
    def size(self) -> int:
        return len(self._set)

    def __len__(self) -> int:
        return len(self._set)

File: proposed-projects/round-trip-experiment/reverse/20260909T154949Z_461ef9c5/20260910T092120Z_46cf829b/ported_implementation/src/grammar-graph/generic-set.ts
export class GenericSet<T, K> {
  private getKey: (el: T) => K;
  private keys = new Map<K, T>();

  constructor(getKey: (el: T) => K) {
    this.getKey = getKey;
  }

  add(el: T): void {
    const key = this.getKey(el);
    if (!this.keys.has(key)) {
      this.keys.set(key, el);
    }
  }

  delete(el: T): void {
    const key = this.getKey(el);
    if (!this.keys.has(key)) {
      throw new Error('Could not get ref');
    }
    this.keys.delete(key);
  }

  has(el: T): boolean {
    return this.keys.has(this.getKey(el));
  }

  get(el: T): T | undefined {
    return this.keys.get(this.getKey(el));
  }

  get size(): number {
    return this.keys.size;
  }

  *[Symbol.iterator](): IterableIterator<T> {
    yield* [...this.keys.values()];
  }
}
